### PR TITLE
BLAKE2 based MACs

### PR DESCRIPTION
**Pull request by karel-m**

### CODE REVIEW

**Comment by karel-m** [comment]
TODO: use test vectors from https://github.com/BLAKE2/BLAKE2

**Comment by sjaeckel** [comment]
can you add the memory functions to multi.c please and rebase? :)

**Comment by sjaeckel** [comment]
> TODO: use test vectors from https://github.com/BLAKE2/BLAKE2

uhm sure, but then please make it optional to run all of them (as already done for the RSA testvectors).

**Comment by karel-m** [comment]
> can you add the memory functions to multi.c

I am not sure what you mean, something has to be fixed in `blake2bmac_memory_multi.c` / `blake2smac_memory_multi.c`?

**Comment by sjaeckel** [comment]
> > can you add the memory functions to multi.c
> 
> I am not sure what you mean, something has to be fixed in `blake2bmac_memory_multi.c` / `blake2smac_memory_multi.c`?

nope, I meant adding the functions to `demos/multi.c`

**Comment by karel-m** [comment]
@sjaeckel what about to turn `demos/multi.c` into `testprof/multi_test.c`?

**Comment by karel-m** [comment]
One maybe important difference related to `blake2[bs]mac_memory` + `blake2[bs]mac_memory_multi` functions - the parameter `unsigned long *maclen` must be set exactly to the required size as is is used not only by `blake2[bs]mac_done` functions but also by `blake2[bs]mac_init` functions.

When calling other MAC's `*_memory` functions you can pass a "longer" values in `unsigned long *maclen` which will be set to the actual value of the output length.

We can add an extra parameter (e.g. `blake2len`) to all `blake2[sb]mac_memory.c` + `blake2[sb]mac_memory_multi.c`

**Comment by sjaeckel** [comment]
> One maybe important difference related to `blake2[bs]mac_memory` + `blake2[bs]mac_memory_multi` functions - the parameter `unsigned long *maclen` must be set exactly to the required size as is is used not only by `blake2[bs]mac_done` functions but also by `blake2[bs]mac_init` functions.
> 
> When calling other MAC's `*_memory` functions you can pass a "longer" values in `unsigned long *maclen` which will be set to the actual value of the output length.
> 
> We can add an extra parameter (e.g. `blake2len`) to all `blake2[sb]mac_memory.c` + `blake2[sb]mac_memory_multi.c`

I think that's perfectly fine! Which reminds me that we just have to write documentation... for all the changes/new features...

**Comment by karel-m** [comment]
So, are we ready for merging here?

File: src/mac/blake2/blake2bmac_file.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  @param fname    The name of the file you wish to BLAKE2B MAC
+  @param key      The secret key
+  @param keylen   The length of the secret key
+  @param mac      [out] The BLAKE2B MAC authentication tag
+  @param maclen   [in/out]  The max size and resulting size of the authentication tag
+  @return CRYPT_OK if successful, CRYPT_NOP if file support has been disabled
+*/
+int blake2bmac_file(const char *fname, const unsigned char *key, unsigned long keylen, unsigned char *mac, unsigned long *maclen)
+{
+        #ifdef LTC_NO_FILE
```
Comment (sjaeckel) [inline src/mac/blake2/blake2bmac_file.c]: indentation

Comment (karel-m) [inline src/mac/blake2/blake2bmac_file.c]: fixed

**Pull request merged**